manager/assets/modext/widgets/source/modx.panel.sources.js

```diff
@@ -63,221 +68,377 @@ Ext.reg('modx-panel-sources',MODx.panel.Sources);
                             MODx.request.query = '';
                         }
-                    }
-                    ,scope: this
-                }
-                ,'render': {
+                    },
+                    scope: this
+                },
+                render: {
                     fn: function(cmp) {
                         new Ext.KeyMap(cmp.getEl(), {
-                            key: Ext.EventObject.ENTER
-                            ,fn: this.blur
-                            ,scope: cmp
+                            key: Ext.EventObject.ENTER,
+                            fn: this.blur,
+                            scope: cmp
                         });
-                    }
-                    ,scope: this
+                    },
+                    scope: this
                 }
             }
-        },{
-            xtype: 'button'
-            ,text: _('filter_clear')
-            ,id: 'modx-filter-clear'
-            ,cls: 'x-form-filter-clear'
-            ,listeners: {
-                'click': {fn: this.clearFilter, scope: this},
-                'mouseout': { fn: function(evt){
-                    this.removeClass('x-btn-focus');
-                }
+        }, {
+            xtype: 'button',
+            text: _('filter_clear'),
+            id: 'modx-filter-clear',
+            cls: 'x-form-filter-clear',
+            listeners: {
+                click: {
+                    fn: this.clearFilter,
+                    scope: this
+                },
+                mouseout: {
+                    fn: function(evt) {
+                        this.removeClass('x-btn-focus');
+                    }
                 }
             }
-        }]
+        }],
+        viewConfig: {
+            forceFit: true,
+            scrollOffset: 0,
+            getRowClass: function(record, index, rowParams, store) {
+                // Adds the returned class to the row container's css classes
+                const disableClasses = record.json.isProtected ? 'modx-protected-row disable-selection' : 'disable-selection' ;
+                return !this.grid.userCanDeleteRecord(record) ? disableClasses : '';
+            }
+        }
+    });
+    MODx.grid.Sources.superclass.constructor.call(this, config);
+
+    this.protectedDataIndex = 'id';
+    this.protectedIdentifiers = [1];
+    this.nonRemoveableRecords = [];
+    this.gridMenuActions = ['edit', 'delete', 'duplicate'];
+
+    this.setUserCanEdit(['source_save', 'source_edit']);
+    this.setUserCanCreate(['source_save']);
+    this.setUserCanDelete(['source_delete']);
+    this.setShowActionsMenu();
+
+    this.on({
+        render: function(grid) {
+            this.setEditableColumnAccess(
+                ['modx-source--name', 'modx-source--description']
+            );
+        },
+        beforeedit: function(e){
+            if (e.record.json.name === 'Filesystem' || !this.userCanEditRecord(e.record)) {
```

Comment:
Hardcode

Reply:
No, I'm pretty sure this has to be done on-the-fly.